pennylane/interfaces/autograd.py

```diff
@@ -67,14 +68,39 @@ def gradient_product(g):
                     nested Sequence[float]: vector-Jacobian product, arranged
                     into the nested structure of the input arguments in ``args``
                 """
+                diff_args = []
+                diff_indices = None
+                non_diff_indices = set()
+
+                for arg, variable_inds in zip(args, self.arg_vars):
+                    if not getattr(arg, "requires_grad", True):
+                        indices = [i.idx for i in _flatten(variable_inds)]
+                        non_diff_indices.update(indices)
+                    else:
+                        diff_args.append(arg)
+
+                if non_diff_indices:
+                    diff_indices = set(range(self.num_variables)) - non_diff_indices
```

Comment:
Is this just to avoid the case that a variable that is used in multiple gates might get double counted?

Reply:
Nope, this logic is independent of the underlying QNode operations. Essentially:

1. On construction, the `BaseQNode` takes the input positional arguments `[x[2, 3], y[4], z[1], ...]` (I've included example shapes in square brackets) and flattens it into a single list of `Variables`, with each one corresponding to an _element_ of the input arrays:

   ```pycon
   >>> print(arg_vars)
   [Var(0, name="x[0, 0]"), Var(1, name="x[0, 1]"), ..., Var(9, name="y[3]"), Var(10, name="z[0]")]
   >>> len(_flatten(args))
   >>> 1
   ```

2. It then 'unflattens' this list to match the shape of `args` for convenience, and stores this as `self.arg_vars`:

   ```pycon
   >>> print(self.arg_vars)
   [array([[Var(0, name="x[0, 0]"), Var(1, name="x[0, 1]"), Var(2, name="x[0, 2]")],
           [Var(3, name="x[1, 0]"), Var(4, name="x[1, 1]"), Var(5, name="x[1, 2]")]]),
    array([Var(6, name="y[0]"), Var(7, name="y[1]]"), Var(8, name="y[2]"), Var(9, name="y[3]")]),
    [Var(10, name="z[0]")]]
    ```

All I'm doing here is getting the variable indices of the differentiable parameters, so that I can pass them to `jacobian()` using the `wrt` argument. *Only variables specified in `wrt` will have the parameter shift rule applied*, and the returned Jacobian will be of size `(len(wrt), len(ans))`.

So in this example, if `x.requires_grad = z.requires_grad = False` and `y.requires_grad = True`, then `wrt = {6, 7, 8, 9}`.

Reply:
Internally, if Variable 6 is then used in multiple gates, the `QubitQNode._pd_analytic()` method will internally apply the product rule.